File: py/LC_435.py
class List(list):
    pass


class Solution:
    def eraseOverlapIntervals(self, intervals: List[List[int]]) -> int:

        intervals.sort()
        prev_end = intervals[0][1]
        res = 0

        for start, end in intervals[1:]:
            if start >= prev_end:
                prev_end = end
                continue
            else:
                res += 1
                prev_end = min(end, prev_end)
        
        return res


def test():
    s = Solution()

    assert s.eraseOverlapIntervals([[1,2],[2,3],[3,4],[1,3]]) == 1
    assert s.eraseOverlapIntervals([[1,2],[1,2],[1,2]]) == 2 
    assert s.eraseOverlapIntervals([[1,2],[2,3]]) == 0 
    assert s.eraseOverlapIntervals([[1,2],[2,3],[3,4],[0,2]]) == 1 
    assert s.eraseOverlapIntervals([[1,2],[2,3],[3,7],[4,7]]) == 1
    assert s.eraseOverlapIntervals([[1,2],[2,3],[3,7],[3,6]]) == 1 
    assert s.eraseOverlapIntervals([[1,2],[2,3],[3,7],[3,7]]) == 1
    assert s.eraseOverlapIntervals([[1,100], [1, 11], [2, 12], [11,22]]) == 2


if __name__ == "__main__":
    test()

File: cpp/LC_435.cpp
#include <iostream>
#include <vector>
#include <algorithm>

class Solution {
public:
    int eraseOverlapIntervals(std::vector<std::vector<int>>& intervals) {
        std::sort(intervals.begin(), intervals.end());

        int ans = 0;
        int prevEnd = intervals[0][1];
        for (int i = 1; i < intervals.size(); i++) {
            if (prevEnd > intervals[i][0]) {
                ans++;
                prevEnd = std::min(prevEnd, intervals[i][1]);
            }
            else {
                prevEnd = intervals[i][1];
            }
        }

        return ans++;
    }
};

int main() {
    Solution s;
    std::vector<std::vector<int>> inp {{1, 2}, {2, 3}, {3, 4}, {1, 3}};
    std::cout << s.eraseOverlapIntervals(inp) << " 1\n";

    inp = {{1, 2}, {1, 2}, {1, 2}};
    std::cout << s.eraseOverlapIntervals(inp) << " 2\n";

    inp = {{1, 2}, {2, 3}};
    std::cout << s.eraseOverlapIntervals(inp) << " 0\n";

    inp = {{2, 3}, {3, 4}, {1, 8}, {5, 6}, {7, 9}, {8, 9}};
    std::cout << s.eraseOverlapIntervals(inp) << " 2\n";
    return 0;
}